File: utility/AppendDataToMat.m
function AppendDataToMat(serverUrl, dataDir)
    % APPENDDATATOMAT Fetches parameters from the server and appends them 
    % to existing .mat files using the native -append feature.
    % 
    % Usage:
    %   AppendDataToMat() 
    %   AppendDataToMat('http://192.168.1.5:3753', 'C:/my/data/dir')

    if nargin < 1
        serverUrl = 'http://127.0.0.1:3753';
    end
    if nargin < 2
        dataDir = fullfile(pwd, '..', 'server', 'data');
    end
    
    fprintf('========================================\n');
    fprintf('Starting AppendDataToMat Utility\n');
    fprintf('Server URL: %%s\n', serverUrl);
    fprintf('Data Directory: %%s\n', dataDir);
    fprintf('========================================\n\n');
    
    %% Fetch batch info from server
    batchUrl = sprintf('%%s/batchInfo', serverUrl);
    fprintf('Fetching batch data from %%s...\n', batchUrl);
    
    options = weboptions('Timeout', 60);
    try
        batchData = webread(batchUrl, options);
    catch ME
        fprintf('ERROR: Failed to fetch batch info from server.\n');
        fprintf('Message: %%s\n', ME.message);
        return;
    end
    
    %% Parse JSON into a fast lookup map
    paramMap = containers.Map('KeyType', 'double', 'ValueType', 'any');
    
    if iscell(batchData)
        numItems = length(batchData);
        for i = 1:numItems
            item = batchData{i};
            if isfield(item, 'id')
                paramMap(item.id) = item;
            elseif isfield(item, 'index')
                paramMap(item.index) = item;
            end
        end
    elseif isstruct(batchData)
        numItems = length(batchData);
        for i = 1:numItems
            item = batchData(i);
            if isfield(item, 'id')
                paramMap(item.id) = item;
            elseif isfield(item, 'index')
                paramMap(item.index) = item;
            end
        end
    else
        fprintf('ERROR: Unrecognized data format received from server.\n');
        return;
    end
    
    fprintf('Successfully fetched and mapped %%d parameter configurations.\n\n', paramMap.Count);
    
    %% Scan the directory for .mat files
    fprintf('Scanning directory %%s for exp-*.mat files...\n', dataDir);
    files = dir(fullfile(dataDir, 'exp-*.mat'));
    if isempty(files)
        fprintf('No exp-*.mat files found in the target directory.\n');
        return;
    end
    
    successCount = 0;
    skipCount = 0;
    errorCount = 0;
    
    fprintf('\n--- Beginning File Modifications ---\n');
    
    %% Process each file
    for i = 1:length(files)
        fileName = files(i).name;
        filePath = fullfile(dataDir, fileName);
        
        fprintf('[%%d/%%d] Processing %%s... ', i, length(files), fileName);
        
        % Extract ID using regex
        tokens = regexp(fileName, 'exp-(\d+)\.mat', 'tokens');
        if isempty(tokens)
            fprintf('SKIPPED (Could not parse ID from filename).\n');
            skipCount = skipCount + 1;
            continue;
        end
        
        fileID = str2double(tokens{1}{1});
        
        if isKey(paramMap, fileID)
            try
                % Get the specific parameter structure for this ID
                data = paramMap(fileID);
                
                % Append to the .mat file (leaves op/algo untouched)
                save(filePath, 'data', '-append');
                
                fprintf('SUCCESS (Appended data for ID %%d).\n', fileID);
                successCount = successCount + 1;
            catch ME
                fprintf('ERROR (%%s).\n', ME.message);
                errorCount = errorCount + 1;
            end
        else
            fprintf('SKIPPED (ID %%d not found in fetched server data).\n', fileID);
            skipCount = skipCount + 1;
        end
    end
    
    %% Summary
    fprintf('\n========================================\n');
    fprintf('Operation Complete Summary:\n');
    fprintf('  Total files scanned: %%d\n', length(files));
    fprintf('  Successfully updated: %%d\n', successCount);
    fprintf('  Skipped (No ID match): %%d\n', skipCount);
    fprintf('  Errors encountered: %%d\n', errorCount);
    fprintf('========================================\n');
end

File: utility/append_data_to_mat.py
import os
import argparse
import json
import urllib.request
import scipy.io


def fetch_batch_info(server_url):
    """
    Fetches the full array of generated parameters from the server.
    """
    print(f"Fetching data from {server_url}/batchInfo...")
    try:
        req = urllib.request.Request(f"{server_url}/batchInfo")
        with urllib.request.urlopen(req, timeout=10) as response:
            data = json.loads(response.read().decode('utf-8'))
    except Exception as e:
        print(f"Failed to connect to the server: {e}")
        return {}
    
    # Build a mapping of id -> parameter dictionary
    param_map = {}
    for item in data:
        # Check for 'id' first, fallback to 'index' if necessary
        item_id = item.get('id', item.get('index'))
        if item_id is not None:
            param_map[int(item_id)] = item
            
    print(f"Fetched {len(param_map)} parameter configurations from the server.")
    return param_map


def append_data_to_mat(server_url, data_dir):
    """
    Scans the data directory for .mat files, matches them to the server data,
    and injects the data variable into the file.
    """
    if not os.path.isdir(data_dir):
        print(f"Error: Directory '{data_dir}' does not exist.")
        return

    param_map = fetch_batch_info(server_url)
    if not param_map:
        print("No parameters found or server unreachable. Exiting.")
        return

    success_count = 0
    error_count = 0
    skip_count = 0

    print(f"\nScanning '{data_dir}' for .mat files...")
    
    for filename in os.listdir(data_dir):
        if filename.startswith("exp-") and filename.endswith(".mat"):
            try:
                # Extract ID from filename (e.g., exp-15.mat -> 15)
                id_str = filename.replace("exp-", "").replace(".mat", "")
                file_id = int(id_str)
            except ValueError:
                continue
            
            if file_id in param_map:
                file_path = os.path.join(data_dir, filename)
                try:
                    # Load existing .mat file
                    mat_data = scipy.io.loadmat(file_path)
                    
                    # Inject the parameter dictionary as the 'data' variable
                    mat_data['data'] = param_map[file_id]
                    
                    # Overwrite the file with the new data included
                    scipy.io.savemat(file_path, mat_data)
                    success_count += 1
                except Exception as e:
                    print(f"Error modifying {filename}: {e}")
                    error_count += 1
            else:
                skip_count += 1

    print(f"\nOperation Complete:")
    print(f"  Successfully updated: {success_count} files")
    print(f"  Skipped (No ID match): {skip_count} files")
    print(f"  Errors encountered: {error_count} files")


def main():
    parser = argparse.ArgumentParser(description="Inject server parameters into corresponding .mat files.")
    parser.add_argument("--host", type=str, default="127.0.0.1", help="Server host IP")
    parser.add_argument("--port", type=int, default=3753, help="Server port")
    
    # Default data directory to ../server/data relative to this script
    script_dir = os.path.dirname(os.path.abspath(__file__))
    default_data = os.path.normpath(os.path.join(script_dir, "..", "server", "data"))
    
    parser.add_argument("--data-dir", type=str, default=default_data, help="Path to the data directory containing .mat files")
    
    args = parser.parse_args()
    server_url = f"http://{args.host}:{args.port}"
    
    append_data_to_mat(server_url, args.data_dir)


if __name__ == "__main__":
    main()
